In [1]:
using Pkg; Pkg.activate(joinpath(Pkg.devdir(), "MLCourse"))
using CSV, DataFrames, Plots, OpenML , MLJ, Serialization

  Activating project at `~/.julia/dev/MLCourse`


# Raw Data Visualisation

Let's start looking at the raw training data:

In [3]:
train_data = CSV.read("DATA/train.csv", DataFrame) 

5000×32286 DataFrame
  Row │ Xkr4     Gm1992   Gm19938   Gm37381  Rp1      Sox17    Gm37587  Gm3732 ⋯
      │ Float64  Float64  Float64   Float64  Float64  Float64  Float64  Float6 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.19038      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
    2 │ 2.86198      0.0  0.506726      0.0      0.0      0.0      0.0      0.
    3 │ 2.76676      0.0  0.629614      0.0      0.0      0.0      0.0      0.
    4 │ 2.14643      0.0  0.0           0.0      0.0      0.0      0.0      0.
    5 │ 2.84005      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
    6 │ 0.0          0.0  0.0           0.0      0.0      0.0      0.0      0.
    7 │ 1.79552      0.0  0.0           0.0      0.0      0.0      0.0      0.
    8 │ 1.96702      0.0  0.0           0.0      0.0      0.0      0.0      0.
  ⋮   │    ⋮        ⋮        ⋮         ⋮        ⋮        ⋮        ⋮        ⋮   ⋱
 4994 │ 2.93918      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
 4995 │ 2.09944      0.0  0.0           0.0      0.0      0.0      0.0      0.
 4996 │ 2.10858      0.0  0.0           0.0      0.0      0.0      0.0      0.
 4997 │ 3.10897      0.0  0.783343      0.0      0.0      0.0      0.0      0.
 4998 │ 2.02595      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
 4999 │ 1.22754      0.0  0.79137       0.0      0.0      0.0      0.0      0.
 5000 │ 2.48507      0.0  0.55012       0.0      0.0      0.0      0.0      0.
                                             32279 columns and 4985 rows omitted

In [4]:
count(train_data.labels.== "KAT5") #returns total number of cells in category "KAT5"

1556

In [5]:
count(train_data.labels.== "eGFP") #returns total number of cells in category "eGFP"

1592

In [6]:

count(train_data.labels.== "CBP")#returns total number of cells in category "CBP"

1852

We can tell from these numbers that: in the training data set, all 3 categories are well represented.

In [7]:
describe(train_data) #returns statistics informations about each predictors

32286×7 DataFrame
   Row │ variable    mean         min  median   max       nmissing  eltype   
       │ Symbol      Union…       Any  Union…   Any       Int64     DataType 
───────┼─────────────────────────────────────────────────────────────────────
     1 │ Xkr4        2.41064      0.0  2.51936  4.26052          0  Float64
     2 │ Gm1992      0.0232695    0.0  0.0      1.8822           0  Float64
     3 │ Gm19938     0.214613     0.0  0.0      1.8791           0  Float64
     4 │ Gm37381     0.000976904  0.0  0.0      1.20868          0  Float64
     5 │ Rp1         0.00707406   0.0  0.0      1.34236          0  Float64
     6 │ Sox17       0.000259006  0.0  0.0      1.29503          0  Float64
     7 │ Gm37587     0.0          0.0  0.0      0.0              0  Float64
     8 │ Gm37323     0.000239274  0.0  0.0      0.721071         0  Float64
   ⋮   │     ⋮            ⋮        ⋮      ⋮        ⋮         ⋮         ⋮
 32280 │ AC124606.2  0.0          0.0  0.0      0.0              0  Float64
 32281 │ AC124606.1  0.0          0.0  0.0      0.0              0  Float64
 32282 │ AC133095.2  0.0          0.0  0.0      0.0              0  Float64
 32283 │ AC133095.1  0.0          0.0  0.0      0.0              0  Float64
 32284 │ AC234645.1  0.0          0.0  0.0      0.0              0  Float64
 32285 │ AC149090.1  1.4093       0.0  1.5325   3.51328          0  Float64
 32286 │ labels                   CBP           eGFP             0  String7
                                                           32271 rows omitted

We can see that some genes have a mean expression of zero, plus a minimal value and a maximal value of 0. This type of parameters are useless to our model prediction. Therefore some data cleaning needs to be done. 

# Data Treatment

## 1-Removal of constant and correlated parameters

In Data_treatment.jl :
- we tested if we had missing data: none found
- then we removed: 
    - the constant parameters
    - the correlated pararameters


In [2]:
train_data_treated = CSV.read("DATA/trainX.csv", DataFrame)

5000×26099 DataFrame
  Row │ Xkr4     Gm1992   Gm19938   Gm37381  Rp1      Sox17    Gm37323  Mrpl15 ⋯
      │ Float64  Float64  Float64   Float64  Float64  Float64  Float64  Float6 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.19038      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
    2 │ 2.86198      0.0  0.506726      0.0      0.0      0.0      0.0      0.
    3 │ 2.76676      0.0  0.629614      0.0      0.0      0.0      0.0      0.
    4 │ 2.14643      0.0  0.0           0.0      0.0      0.0      0.0      0.
    5 │ 2.84005      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
    6 │ 0.0          0.0  0.0           0.0      0.0      0.0      0.0      0.
    7 │ 1.79552      0.0  0.0           0.0      0.0      0.0      0.0      0.
    8 │ 1.96702      0.0  0.0           0.0      0.0      0.0      0.0      0.
  ⋮   │    ⋮        ⋮        ⋮         ⋮        ⋮        ⋮        ⋮        ⋮   ⋱
 4994 │ 2.93918      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
 4995 │ 2.09944      0.0  0.0           0.0      0.0      0.0      0.0      0.
 4996 │ 2.10858      0.0  0.0           0.0      0.0      0.0      0.0      0.
 4997 │ 3.10897      0.0  0.783343      0.0      0.0      0.0      0.0      0.
 4998 │ 2.02595      0.0  0.0           0.0      0.0      0.0      0.0      0. ⋯
 4999 │ 1.22754      0.0  0.79137       0.0      0.0      0.0      0.0      0.
 5000 │ 2.48507      0.0  0.55012       0.0      0.0      0.0      0.0      0.
                                             26092 columns and 4985 rows omitted

In [9]:
describe(train_data_treated)

26099×7 DataFrame
   Row │ variable        mean         min      median   max       nmissing  el ⋯
       │ Symbol          Float64      Float64  Float64  Float64   Int64     Da ⋯
───────┼────────────────────────────────────────────────────────────────────────
     1 │ Xkr4            2.41064          0.0  2.51936  4.26052          0  Fl ⋯
     2 │ Gm1992          0.0232695        0.0  0.0      1.8822           0  Fl
     3 │ Gm19938         0.214613         0.0  0.0      1.8791           0  Fl
     4 │ Gm37381         0.000976904      0.0  0.0      1.20868          0  Fl
     5 │ Rp1             0.00707406       0.0  0.0      1.34236          0  Fl ⋯
     6 │ Sox17           0.000259006      0.0  0.0      1.29503          0  Fl
     7 │ Gm37323         0.000239274      0.0  0.0      0.721071         0  Fl
     8 │ Mrpl15          0.133054         0.0  0.0      2.3545           0  Fl
   ⋮   │       ⋮              ⋮          ⋮        ⋮        ⋮         ⋮         ⋱
 26093 │ Spry3           0.012543         0.0  0.0      1.31561          0  Fl ⋯
 26094 │ Tmlhe           0.291208         0.0  0.0      2.67508          0  Fl
 26095 │ Vmn2r122        8.87452e-5       0.0  0.0      0.443726         0  Fl
 26096 │ 4933409K07Rik   0.0180418        0.0  0.0      1.59582          0  Fl
 26097 │ Gm10931         0.00306767       0.0  0.0      1.05147          0  Fl ⋯
 26098 │ CAAA01147332.1  0.000355018      0.0  0.0      0.580659         0  Fl
 26099 │ AC149090.1      1.4093           0.0  1.5325   3.51328          0  Fl
                                                 1 column and 26084 rows omitted

All the parametors with a 0 mean are gone.
However we can still notice that there might be outliers in our data sets. For example for Gm37323(row 7) the mean is 0.000239274 but the maximal value is 0.721071. 
 

In [10]:
using Statistics
t = train_data_treated.Gm37323 #takes all data points in the column "Gm37323"
u = quantile(t, 0.99) #returns 0.99 quantile 
a = t[t.> u] # returns expression of the cells wich value is above the 0.99 quantile for Gm37323 expression

0.012212418396005238

Thus 2 data points are above the 0.99 quantile

In [13]:
average = (a[1]+a[2])/5000 

0.000239274035926351

The mean observed in column Gm37323 (see describe aboce) is only due to this 2 observations : only 2 cells express this gene!!!
Data needs further treatment!

## 2- Principal Component Analysis

To further treat our data, we decided to PCA

![img](PLOTS/PCA_2D.png)

![Image](PLOTS/PCA_2D_sd.png)

![Image](PLOTS/TNse.png)

![Image](PLOTS/Pvar_explained.png)

![Image](PLOTS/Biplot_PCA_standardized_1.png)

# Looking at Clusters